core/controllers/base.py

```diff
@@ -178,13 +192,14 @@ def __init__(self, request, response):  # pylint: disable=super-init-not-called
         self.partially_logged_in = False
         self.user_is_scheduled_for_deletion = False
         self.current_user_is_super_admin = False
-        # Once the attribute `normalized_request` is type annotated here, make
-        # sure to fix all the subclasses using normalized_request.get() method
-        # by removing their type: ignore[union-attr] and using a type cast
-        # instead to eliminate the possibility on union types.
-        # e.g. ClassroomAccessValidationHandler.
-        self.normalized_request = None
-        self.normalized_payload = None
+        # Here we use type Any because dict 'self.normalized_request' can
+        # contain normalized version of arg's value, and these arg values
+        # can be of any type. So, to allow every type, we used Any here.
+        self.normalized_request: Optional[Dict[str, Any]] = None
+        # Here we use type Any because dict 'self.normalized_payload' can
+        # contain normalized version of arg's value, and these arg values
+        # can be of any type. So, to allow every type, we used Any here.
+        self.normalized_payload: Optional[Dict[str, Any]] = None
```

Comment:
I don't think generics really needed here because:
1. we still need to define typedDict in every subclass.
2. In some subclasses, `normalized_request` posses different typedDict than `normalized_payload`. So, if we use generics here then both of them is assigned with same type.
3. For optional thing we have to use assert everytime. For example:
```python
class BaseHandler(Generic[T]):
   self.normalized_payload: Optional[T] = None

class BlogAdminHandlerNormalizedPayloadDict(TypedDict):
    """Dict representation of BlogAdminHandler's normalized_payload
    dictionary.
    """

    action: str
    config_property_id: Optional[str]


class BlogAdminHandler(base.BaseHandler)[BlogAdminHandlerNormalizedPayloadDict]:
    """Handler for the blog admin page."""

    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
    URL_PATH_ARGS_SCHEMAS: Dict[str, str] = {}
    HANDLER_ARGS_SCHEMAS = {
        'GET': {},
        'POST': {
            'action': {
                'schema': {
                    'type': 'basestring',
                    'choices': [
                        'save_config_properties', 'revert_config_property']
                }
            },
            'config_property_id': {
                'schema': {
                    'type': 'basestring',
                },
                'default_value': None
            },
        }
    }

    @acl_decorators.can_access_blog_admin_page
    def post(self) -> None:
        """Handles POST requests."""
        assert `self.normalized_payload` is not None   # This is because iternally self.normalized is initialziled as None. And before reaching this point we have already raised error for None schemas.
```

Reply:
1. That is fine. But if we use generic the typing will be enforced more strongly.
2. You can introduce two type variables in the `Generic[X, Y]` and thus add a separate typing for `normalized_request` and for `normalized_payload`.
3. I think assert is fine as we will be sure that it is not `None`.

Reply:
I was working on a Generic approach and found a scenario that does not fit into this generic approach.
Basically, If I'm not wrong different call methods can have different `self.normalized_request` dicts like here in `AdminRoleHandler`  https://github.com/oppia/oppia/blob/29059ed45efe10f87324ff6edbeb8be937d514b3/core/controllers/admin.py#L809 -- 
In this handler, the get method can have self.normalized_request with role and username as Optional[str] whereas in its delete method self.normalized_request can have role and username as only string. But in the Generic approach, we are defining self.normalized_request and self.normalized_payload only once so there is a contradiction in this approach. Can you please take a look at this? @vojtechjelinek @aasiffaizal

Reply:
Generics Added PTAL!